stdlib/join/join_fn.go

```diff
@@ -266,6 +266,11 @@ func defaultRow(key flux.GroupKey, objType semantic.MonoType) values.Object {
 }
 
 func getObjectType(arg *semantic.Argument, cols []flux.ColMeta) (semantic.MonoType, error) {
+	// XXX(sean): This appears to fix the empty table bug, but what's the point of the `getObjectType()`
+	// function if we have this information already? Is this type info not always available?
+	if len(cols) < 1 {
+		return arg.TypeOf()
```

Comment:
I share your concern, its seems backwards that we need to know the output type before we have evaluated any of the output rows. I know there are cases where the semantic type is incomplete, i.e. there are type variables. What happens to the join code in those cases? For example use a `{l with ...}` in the join `as` function so that the record is unbounded. 

Generally seems like we need this code to be more robust to unknown types and we should change the paradigm to just use the actual output type of the rows.  

As an aside hopefully this is the kind of thing that becomes much more straightforward with an IR and VM design.

Reply:
Updated the integration test in such a way that (I think) both the output record and the right input would have incomplete type information. It passes, but I'm still suspicious of it...

Reply:
I think there's generally a problem where sometimes type information is available and sometimes it isn't and I'm not always sure why it sometimes is and sometimes isn't. We've generally had to abstain from relying on the information returned from type inference because of this.

I'd really like to be able to answer this question better but this section of code has really gone through too many hack revisions for me to remember the answer or know it with certainty.

I might try to dive into this and ensure that it's addressed in the VM proposal or I at least understand the reasoning behind it.

Reply:
Gotcha. I'll try to do some manual validation against some real data to see if this fix holds up.

Reply:
Yup. Was able to create the following panic by running a query similar to the testcase, but using `from()` instead of `array.from()` or `gen.tables()`

```
panic: unknown builder for type: invalid
```